File: SpringBoot/src/main/java/com/frontend/controllers/MockController.java
package com.frontend.controllers;

import com.frontend.Global;
import com.frontend.mockable.MCMockGroup;
import com.frontend.mockable.MCMockInfo;
import com.frontend.jsonutil.JSON;
import com.frontend.mappers.MockMapper;
import com.frontend.mockable.MCMockParam;
import com.frontend.mockable.MCMockScene;
import com.frontend.models.MCPagination;
import com.frontend.models.MCResult;
import com.frontend.utils.MybatisError;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.jdbc.UncategorizedSQLException;
import org.springframework.web.bind.annotation.*;
import org.sqlite.SQLiteErrorCode;
import org.sqlite.SQLiteException;

import javax.servlet.http.HttpServletRequest;
import javax.servlet.http.HttpServletResponse;
import java.io.UnsupportedEncodingException;
import java.net.URLEncoder;
import java.util.List;

@RestController
@RequestMapping("/mock")
public class MockController {
  @Autowired
  private MockMapper mockMapper;

  @GetMapping("/app/whole")
  @ResponseBody
  @JSON(type = MCMockInfo.class, include = "id,name,path,enabled,sceneid,scenes")
  @JSON(type = MCMockScene.class, include = "id,name,params")
  @JSON(type = MCMockParam.class, include = "id,name,value")
  public MCResult appWhole() {
    List<MCMockGroup> groups = mockMapper.findAllGroup(Global.getAppId(), Global.getUser().getId());
    for (MCMockGroup group : groups) {
      List<MCMockInfo> mocks = mockMapper.findAllMock(Global.getUser(), group.getId(), new MCPagination(1, 1000));
      mocks.forEach(m -> {
        m.setScenes(mockMapper.findAllScene(m.getId()));
        m.getScenes().forEach(s -> {
            MCMockParam param = new MCMockParam();
            param.setSceneid(s.getId());
            s.setParams(mockMapper.findAllParam(param));
          }
        );
        MCMockScene scene = new MCMockScene();
        scene.setId(0);
        scene.setName("自动");
        scene.setMockid(m.getId());
        m.getScenes().add(0, scene);
      });
      group.setMocks(mocks);
    }
    return MCResult.Success(groups);
  }

  @GetMapping("/list")
  public MCResult mockList(Integer groupid, Integer pageSize, Integer current) {
    MCResult result = MCResult.Success(mockMapper.findAllMock(Global.getUser(), groupid, new MCPagination(current, pageSize)));
    return result;
  }

  @PostMapping("/update")
  public MCResult update(@RequestBody MCMockInfo mock) {
    try {
      if (mock.getId() == null) {
        mockMapper.insertNew(mock);
      } else {
        mockMapper.update(mock);
      }
      Global.update();
      return MCResult.Success();
    } catch (UncategorizedSQLException e) {
      SQLiteException se = (SQLiteException) e.getCause();
      if (se.getResultCode() == SQLiteErrorCode.SQLITE_CONSTRAINT_UNIQUE) {
        return MCResult.Failed(MybatisError.DuplicateEntry);
      }
      return MCResult.Failed(MybatisError.InsertFaield);
    }
  }

  @PostMapping("/group/update")
  public MCResult updateGroup(@RequestBody MCMockGroup group) {
    try {
      group.setUid(Global.getUser().getId());
      mockMapper.updateGroup(group);
      Global.update();
      return MCResult.Success();
    } catch (UncategorizedSQLException e) {
      SQLiteException se = (SQLiteException) e.getCause();
      if (se.getResultCode() == SQLiteErrorCode.SQLITE_CONSTRAINT_UNIQUE) {
        return MCResult.Failed(MybatisError.DuplicateEntry);
      }
      return MCResult.Failed(MybatisError.InsertFaield);
    }
  }

  @GetMapping("/group/list")
  public MCResult groupList() {
    MCResult result = MCResult.Success(mockMapper.findAllGroup(Global.getAppId(), Global.getUser().getId()));
    return result;
  }

  @PostMapping("/group/delete")
  public MCResult deleteGroup(@RequestBody MCMockGroup group) {
    try {
      mockMapper.deleteGroup(group);
      return MCResult.Success();
    } catch (UncategorizedSQLException e) {
      SQLiteException se = (SQLiteException) e.getCause();
      if (se.getResultCode() == SQLiteErrorCode.SQLITE_CONSTRAINT_UNIQUE) {
        return MCResult.Failed(MybatisError.DuplicateEntry);
      }
      return MCResult.Failed(MybatisError.InsertFaield);
    }
  }

  @PostMapping("/delete/{id}")
  public MCResult deleteMock(@PathVariable("id") Integer id) {
    mockMapper.deleteMock(id);
    return MCResult.Success();
  }

  @RequestMapping(value = "/app/scene/{id}", produces = "application/json; charset=utf-8", method = {RequestMethod.GET, RequestMethod.POST})
  public String appScene(@PathVariable("id") Integer id, HttpServletRequest request, HttpServletResponse response) throws UnsupportedEncodingException {
    MCMockScene scene = mockMapper.findScene(id);
    if (scene == null) {
      return "scene id " + id + " not found.";
    } else {
      response.setHeader("scene_id", scene.getId().toString());
      response.setHeader("scene_name", URLEncoder.encode(scene.getName(), "utf-8"));
      return scene.getResponse();
    }
  }

  @GetMapping("/scene/list")
  public MCResult sceneList(Integer mockid) {
    List<MCMockScene> scenes = mockMapper.findAllScene(mockid);
    for (MCMockScene scene : scenes) {
      MCMockParam param = new MCMockParam();
      param.setSceneid(scene.getId());
      scene.setParams(mockMapper.findAllParam(param));
      if (scene.getParams().size() == 0) {
        scene.getParams().add(param);
      }
    }
    MCResult result = MCResult.Success(scenes);
    return result;
  }

  @PostMapping("/scene/update")
  public MCResult updateScene(@RequestBody MCMockScene scene) {
    try {
      mockMapper.updateScene(scene);
      return MCResult.Success();
    } catch (UncategorizedSQLException e) {
      SQLiteException se = (SQLiteException) e.getCause();
      if (se.getResultCode() == SQLiteErrorCode.SQLITE_CONSTRAINT_UNIQUE) {
        return MCResult.Failed(MybatisError.DuplicateEntry);
      }
      return MCResult.Failed(MybatisError.InsertFaield);
    }
  }

  @PostMapping("/active")
  public MCResult activeMock(@RequestBody MCMockInfo mock) {
    mockMapper.active(mock);
    Global.update();
    return MCResult.Success();
  }

  @PostMapping("/scene/active")
  public MCResult activeScene(@RequestBody MCMockInfo mock) {
    mockMapper.activeScene(mock);
    Global.update();
    return MCResult.Success();
  }

  @PostMapping("/scene/delete")
  public MCResult deleteScene(@RequestBody MCMockScene scene) {
    try {
      mockMapper.deleteScene(scene.getId());
      return MCResult.Success();
    } catch (UncategorizedSQLException e) {
      SQLiteException se = (SQLiteException) e.getCause();
      if (se.getResultCode() == SQLiteErrorCode.SQLITE_CONSTRAINT_UNIQUE) {
        return MCResult.Failed(MybatisError.DuplicateEntry);
      }
      return MCResult.Failed(MybatisError.InsertFaield);
    }
  }

  @GetMapping("/param/list")
  public MCResult paramList(MCMockParam param) {
    MCResult result = MCResult.Success(mockMapper.findAllParam(param));
    return result;
  }

  @PostMapping("/param/update")
  public MCResult paramUpdate(@RequestBody MCMockParam param) {
    try {
      mockMapper.updateParam(param);
      return MCResult.Success();
    } catch (UncategorizedSQLException e) {
      SQLiteException se = (SQLiteException) e.getCause();
      if (se.getResultCode() == SQLiteErrorCode.SQLITE_CONSTRAINT_UNIQUE) {
        return MCResult.Failed(MybatisError.DuplicateEntry);
      }
      return MCResult.Failed(MybatisError.InsertFaield);
    }
  }

  @PostMapping("/param/delete")
  public MCResult paramDelete(@RequestBody MCMockParam param) {
    try {
      mockMapper.deleteParam(param.getId());
      return MCResult.Success();
    } catch (UncategorizedSQLException e) {
      SQLiteException se = (SQLiteException) e.getCause();
      if (se.getResultCode() == SQLiteErrorCode.SQLITE_CONSTRAINT_UNIQUE) {
        return MCResult.Failed(MybatisError.DuplicateEntry);
      }
      return MCResult.Failed(MybatisError.InsertFaield);
    }
  }

}

File: backend/Sources/CanaryBackend/Controller/MockController.swift
//
//  MockController.swift
//  
//
//  Created by Rake Yang on 2021/6/23.
//

import Foundation
import CanaryProto
import SwiftyJSON

class MockController {
    @Mapping(path: "/mock/app/whole")
    var appWhole: ResultHandler = { req, res in
        if var groups = try MockMapper.findAllGroup(pid: req.pid, uid: req.uid)?.decode([ProtoMockGroup].self) {
            for (index, item) in groups.enumerated() {
                if var mocks = try MockMapper.findAllMock(pid: req.pid, uid: req.uid, paging: Paging(["pageSize": "1000", "pageNum": "1"]), groupId: item.id)?.decode([ProtoMock].self) {
                    for (i, mock) in mocks.enumerated() {
                        mocks[i].scenes = try MockMapper.findAllScene(mockid: mock.id)?.decode([ProtoMockScene].self)
                        for (y, scene) in (mocks[i].scenes ?? []).enumerated() {
                            mocks[i].scenes?[y].params = try MockMapper.findAllParam(sceneid: scene.id)?.decode([ProtoMockParam].self)
                        }
                        mocks[i].scenes?.insert(ProtoMockScene(mockid: mock.id, name: "自动"), at: 0)
                    }
                    groups[index].mocks = mocks
                }
            }
            return .entry(try JSONEncoder().encode(groups))
        }
        return .done
    }
    
    @Mapping(path: "/mock/list")
    var mockList: ResultHandler = { request, response in
        let paging = Paging(request.getDictionary)
        return .entry(try MockMapper.findAllMock(pid: request.pid, uid: request.uid, paging: paging, groupId: request.getDictionary.intValue("groupid")))
    }
    
    @Mapping(path: "/mock/update", method: [.post])
    var updateMock: ResultHandler = { request, response in
        let mock = try request.postDictionary.decode(ProtoMock.self)
        try MockMapper.update(mock: mock)
        return .done
    }
    
    @Mapping(path: "/mock/active", method: [.post])
    var activeMock: ResultHandler = { req, res in
        try MockMapper.activeMock(mockid: req.postDictionary.intValue("mockid"), enabled: req.postDictionary.boolValue("enabled"))
        return .done
    }
    
    @Mapping(path: "/mock/delete/{id}", method: [.post])
    var deleteMock: ResultHandler = {  request, response in
        try MockMapper.deleteMock(mockid: request.urlVariables.intValue("id"))
        return .done
    }
    
    @Mapping(path: "/mock/group/list")
    var groupList: ResultHandler = { request, response in
        return .entry(try MockMapper.findAllGroup(pid: request.pid, uid: request.uid))
    }
    
    @Mapping(path: "/mock/group/update", method: [.post])
    var groupUpdate: ResultHandler = { request, response in
        try MockMapper.updateGroup(request.postDictionary.decode(ProtoMockGroup.self))
        return .done
    }
    
    @Mapping(path: "/mock/group/delete/{id}", method: [.post])
    var groupDelete: ResultHandler = { request, resposne in
        try MockMapper.deleteGroup(request.urlVariables.intValue("id"))
        return .done
    }
    
    @Mapping(path: "/mock/scene/list/{mockid}")
    var sceneList: ResultHandler = { request, response in
        var scenes = try MockMapper.findAllScene(mockid: request.urlVariables.intValue("mockid"))?.decode([ProtoMockScene].self)
        for (index, item) in (scenes ?? []).enumerated() {
            scenes?[index].params = try MockMapper.findAllParam(sceneid: item.id)?.decode([ProtoMockParam].self)
        }
        return .entry(try JSONEncoder().encode(scenes))
    }
    
    @Mapping(path: "/mock/app/scene/{id}", method: [.get, .post])
    var sceneDetail: ResultHandler = { request, response in
        let scene = try MockMapper.findScene(sceneId: request.urlVariables.intValue("id"))?.decode(ProtoMockScene.self)
        if let scene = scene {
            response.setHeader(.contentType, value: "application/json; charset=utf-8")
            response.setHeader(.custom(name: "Scene-Id"), value: String(scene.id))
            response.setHeader(.custom(name: "Scene-Name"), value: scene.name.stringByEncodingURL)
            try response.setBody(json: scene.response.jsonDecode())
            response.completed(status: .ok)
        } else {
            response.completed(status: .notFound)
        }
        return nil
    }
    
    @Mapping(path: "/mock/scene/update", method: [.post])
    var updateScene: ResultHandler = { request, response in
        try MockMapper.updateScene(scene: request.postDictionary.decode(ProtoMockScene.self))
        return .done
    }
    
    @Mapping(path: "/mock/param/update", method: [.post])
    var updateParam: ResultHandler = { request, response in
        try MockMapper.updateParam(param: request.postDictionary.decode(ProtoMockParam.self))
        return .done
    }
    
    @Mapping(path: "/mock/param/delete/{id}", method: [.post])
    var deleteParam: ResultHandler = { request, response in
        try MockMapper.deleteParam(paramid: request.urlVariables.intValue("id"))
        return .done
    }
    
    @Mapping(path: "/mock/scene/delete/{id}", method: [.post])
    var deleteScene: ResultHandler = {  request, response in
        try MockMapper.deleteScene(sceneid: request.urlVariables.intValue("id"))
        return .done
    }
    
    @Mapping(path: "/mock/scene/active", method: [.post])
    var activeScene: ResultHandler = { request, response in
        try MockMapper.activeScene(request.postDictionary.intValue("sceneid"), enabled: request.postDictionary.boolValue("enabled"), mockid: request.postDictionary.intValue("mockid"))
        return .done
    }
}
